### PR TITLE
Fix #1196 unmatched extern "C" brace

### LINKED ISSUE #1196: extern "C" { } not matched
found in latest release 0.65.

debug file: test.h
```cpp
#ifdef __cplusplus
extern "C" {
#endif


void test(void);

#ifdef __cplusplus
} // extern "C"
#endif
```
run uncrustify,
```sh
$ uncrustify --version
Uncrustify-0.65
$ uncrustify -c forUncrustifySources.cfg test.h
Output suffix: .uncrustify
Parsing: test.h as language CPP
Unmatched BRACE_CLOSE
at line=9, column=1
```
current version: uncrustify 0.65
working version: uncrustify 0.64

### PR DESCRIPTION
**Pull request by pkolbus**
A common pattern for headers that are used in both C and C++ sources
is to conditionally wrap declarations that need C linkage in an
extern C block, but guarded by preprocessor. Since uncrustify's
parser doesn't yet match braces across preprocessor blocks, the brace
close for this "extern C" was incorrectly seen as unmatched, resulting
in fatal error.

This patch suppresses this fatal error when inside a PP block, until
the parser can be improved.

### CODE REVIEW

File: src/brace_cleanup.cpp [hunk 1]
```diff
@@ -691,7 +691,7 @@ static void parse_cleanup(parse_frame_t *frm, chunk_t *pc)
       LOG_FMT(LSTMT, "%s(%d): %zu> reset expr on %s\n",
               __func__, __LINE__, pc->orig_line, pc->text());
    }
-   else if (pc->type == CT_BRACE_CLOSE)
+   else if (pc->type == CT_BRACE_CLOSE && pc->pp_level == 0)
```
Comment (mihaipopescu) [inline src/brace_cleanup.cpp]: Does this mean that this test would not fail as expected ?
```cpp
#ifndef _INCLUDE_H_
#define _INCLUDE_H_

class Class
{
;
#endif
```

**Comment by gmaurel** [comment]
We need some more tests.

File: src/brace_cleanup.cpp (hunk 1 above)
Comment (pkolbus) [inline src/brace_cleanup.cpp]: Yes, that header no longer fails. What's the best way through this-- check and adapt for ifdef_over_whole_file() ?

On the other hand, that code will not compile... maybe it's easier and better to let the compiler catch such a problem and not leave as defined behavior how Uncrustify will handle such an input? If so then the whole error message could be removed.

**Comment by pkolbus** [comment]
@guy-maurel, happy to add more tests but none come to mind at the moment. Do you have specific cases that you see as missing?

Also-- as far as I can tell, tests/run_tests.py doesn't have provision for tests that are expected to fail... is there another test runner that does, or should I also update run_tests.py?

File: src/brace_cleanup.cpp (hunk 1 above)
Comment (mihaipopescu) [inline src/brace_cleanup.cpp]: There is a test in `scripts` folder that checks if an unmatched brace error is given. You can extend that. If that would have contained the test above this patch would introduce a regression which is not ok.

Your argument about that it's ok that the code will not compile doesn't really hold because you're fixing another case while breaking another.

**Comment by CDanU** [comment]
I see it this way: Issue #1196 is quiet the showstopper for most C/C++ projects. They don't even need to have a bug in their code to run into this.
To fix that issue we will, most likely, have to completely change how the algo that handles the frames is working. Meaning that we won't have a optimal fix soon.

We can handle which bug we want to have: One that makes Uncrustify useless for most projects, or one that makes Uncrustify useless for buggy projects.

Now there is the matter of urgency:
If we want to not break the release cycle we can try to solve this properly, i.e. we won't merge this PR. Most people simply won't care to compile Uncrustify and will sit on the previous version.
The other alternative will be to merge this PR (including the less severe bug) and make a new Release.
Here I would decide to rather take option 1. In this cause we should maybe notify distros, that ship Uncrustify, to skip this release.

**Comment by mihaipopescu** [comment]
I don't understand. In both cases we're dealing with code that doesn't compile, whether is guarded by some ifdefs or not. And I think the usecase I presented is quite common practice. What if you have:

```cpp
#ifndef _HEADER_INCLUDED_
#define _HEADER_INCLUDED_

#ifdef __cplusplus
extern "C" {
#endif

void foo(void);

#ifdef __cplusplus
}
#endif
#endif
```

**Comment by CDanU** [comment]
The fix is for the issue #1196 where Uncrustify fails because it gets a open brace in pp-block N but forgets about it in pp-block M > N so that it appears that the closing brace has no matching opening brace. 

Your first example, on its own, indeed won't compile but the second should.
Am I missing something?

**Comment by mihaipopescu** [comment]
Maybe i dont get it. What error where we showing before and how? Why are we counting brace level in pp?

**Comment by CDanU** [comment]
Inside #1196 you can see the current error message that we are throwing. We don't count the brace level in pp, but every chunk has a pp_level to keep track in how much pp blocks it is nested. Out of the blue I don't know why this is done.

In Uncrustify 0.64 checking if closing braces had opening braces was different. I think we had it in two places.  In 0.65 the checks where moved to brace_cleanup.cpp, that made the checks more strict that before, which is now causing the problems.

**Comment by mihaipopescu** [comment]
I agree we should do something to stop this error being erroneously reported on good code (if I understand correctly) but it looks like that the brace level is incremented in the pp blocks but not decrememted or something. And the fix looks like it's dealing with the effects and not really the cause.

**Comment by mihaipopescu** [comment]
This is really bad. I'm seeing the regression on our code-base as well.

File: src/brace_cleanup.cpp (hunk 1 above)
Comment (pkolbus) [inline src/brace_cleanup.cpp]: Correction: looking closer at the example you provide, that's an unmatched brace open. But I've added a test for unmatched brace close -- basically I-842 but wrapped in a whole file PP. And fixed my breakage.

**Comment by pkolbus** [comment]
Thanks for the pointers.
FWIW, I'm in agreement that this isn't attacking the root cause, just a symptom... but it does unbreak uncrustify for a common use (mixed C/C++ projects) and enables those who want the latest improvements to build and use master again. I'm happy to keep poking at the parser code and replace this code change with something better in a future PR... unless someone else gets there first.

**Comment by mihaipopescu** [comment]
No it's fine. Great work on last fix.

File: src/brace_cleanup.cpp [hunk 2]
```diff
@@ -695,7 +696,8 @@ static void parse_cleanup(parse_frame_t *frm, chunk_t *pc)
    {
       if (!cpd.consumed)
       {
-         if (!cpd.unc_off_used)
+         size_t file_pp_level = ifdef_over_whole_file() ? 1 : 0;
+         if (!cpd.unc_off_used && pc->pp_level == file_pp_level)
```
Comment (mihaipopescu) [inline src/brace_cleanup.cpp]: While this is not ideal is a lot better than before as it covers two most used patterns.
